flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -262,36 +261,92 @@ public void grantLeadership(UUID leaderSessionID) {
+                                                                "verify job result entry",
+                                                                leaderSessionId)))
+                        .handle(
+                                (hasJobResult, throwable) -> {
+                                    if (throwable != null) {
+                                        ExceptionUtils.rethrow(throwable);
+                                    }
+                                    if (hasJobResult == null) {
+                                        return null;
+                                    }
+                                    if (hasJobResult) {
+                                        handleJobAlreadyDoneIfValidLeader(leaderSessionId);
+                                    } else {
+                                        createNewJobMasterServiceProcessIfValidLeader(
+                                                leaderSessionId);
+                                    }
+                                    return null;
+                                });
         handleAsyncOperationError(sequentialOperation, "Could not start the job manager.");
     }
 
-    @GuardedBy("lock")
-    private void verifyJobSchedulingStatusAndCreateJobMasterServiceProcess(UUID leaderSessionId)
-            throws FlinkException {
-        try {
-            if (jobResultStore.hasJobResultEntry(getJobID())) {
-                jobAlreadyDone(leaderSessionId);
+    private <T> CompletableFuture<T> supplyAsyncIfValidLeader(
+            UUID expectedLeaderId,
+            Supplier<CompletableFuture<T>> supplier,
+            Supplier<CompletableFuture<T>> noLeaderFallback) {
+        final CompletableFuture<T> resultFuture = new CompletableFuture<>();
+        runIfValidLeader(
+                expectedLeaderId,
+                () -> FutureUtils.forward(supplier.get(), resultFuture),
+                () -> FutureUtils.forward(noLeaderFallback.get(), resultFuture));
+
+        return resultFuture;
+    }
+
+    private void handleJobAlreadyDoneIfValidLeader(UUID leaderSessionId) {
+        runIfValidLeader(
+                leaderSessionId,
+                () -> {
+                    jobAlreadyDone(leaderSessionId);
+                },
+                () -> printLogIfNotValidLeader("check completed job", leaderSessionId));
+    }
+
+    private void createNewJobMasterServiceProcessIfValidLeader(UUID leaderSessionId) {
+        runIfValidLeader(
+                leaderSessionId,
+                () -> {
+                    ThrowingRunnable.unchecked(
+                                    () -> createNewJobMasterServiceProcess(leaderSessionId))
+                            .run();
+                },
+                () ->
+                        printLogIfNotValidLeader(
+                                "create new job master service process", leaderSessionId));
+    }
+
+    private void runIfValidLeader(
```

Comment:
Moving the `runIfValidLeader` method up destroys the git history to some degree (imagine someone wants to see the history of this specific method in Intellij by selecting the method and using the "Git > Show history for Selection..." feature). Instead, why not refactor the method in its original location and add the `supplyAsyncIfValidLeader` next to that one.

Reply:
1. The `runIfValidLeader ` has been modified in its original location.
2. The `supplyAsyncIfValidLeader ` has been added next to `runIfValidLeader`.